### PR TITLE
Add Kafka and Redis implementation

### PR DESCRIPTION
**Pull request by SlavikN14**
Add Redis and Kafka implementation and related fixes.
Add @DmitryOnopriienko for review.

### CODE REVIEW

File: telegram-bot/src/main/resources/application-docker.yaml [hunk 1]
```diff
@@ -1,7 +1,23 @@
...
+    properties:
+      schema:
+        registry:
+          url: http://schema-registry:8081
+
 nats:
   url: nats://nats:4222
+
+redis-properties:
+  ttl:minutes: 5
```
Comment (ajax-gnatiuk-s) [inline telegram-bot/src/main/resources/application-docker.yaml]:
```suggestion
  ttl:
      minutes: 5
```

File: telegram-bot/src/main/kotlin/com/ajaxproject/telegrambot/service/telegram/TelegramUpdateDispatcherService.kt [hunk 2]
```diff
@@ -46,7 +49,18 @@ class FinanceBot(
         )
     }
 
+    private fun dispatch(updateRequest: UpdateRequest): Mono<Boolean> {
+        return Flux.fromIterable(handlers)
+            .filter { it.isApplicable(updateRequest) }
+            .flatMap {
+                it.handle(updateRequest)
+                    .then(Mono.just(true))
```
Comment (ajax-gnatiuk-s) [inline telegram-bot/src/main/kotlin/com/ajaxproject/telegrambot/service/telegram/TelegramUpdateDispatcherService.kt]:
```suggestion
                    .thenReturn(true)
```

File: telegram-bot/src/main/kotlin/com/ajaxproject/telegrambot/service/CurrencyExchangeService.kt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
+
+@Component
+class CurrencyExchangeService(
+    private val currencyExchangeCacheableRepository: CacheableRepository<MongoCurrency>,
+    private val kafkaProducer: CurrencyKafkaProducer,
+) {
+
+    fun addAllCurrency(arrayCurrency: Array<CurrencyExchangeResponse>): Mono<Unit> {
```
Comment (ajax-gnatiuk-s) [inline telegram-bot/src/main/kotlin/com/ajaxproject/telegrambot/service/CurrencyExchangeService.kt]:
```suggestion
    fun addAllCurrency(currencies: Array<CurrencyExchangeResponse>): Mono<Unit> {
```

File: telegram-bot/src/main/kotlin/com/ajaxproject/telegrambot/service/CurrencyExchangeService.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        return Flux.fromArray(arrayCurrency)
+            .map { it.toEntity() }
+            .filter { checkCurrency(it) }
+            .flatMap { currencyExchangeCacheableRepository.save(it) }
+            .next()
+            .doMonoOnNext { kafkaProducer.sendCurrencyUpdatedEventToKafka(it) }
+            .thenReturn(Unit)
+    }
+
+    fun getCurrencyByCode(code: Int): Mono<List<MongoCurrency>> {
```
Comment (ajax-gnatiuk-s) [inline telegram-bot/src/main/kotlin/com/ajaxproject/telegrambot/service/CurrencyExchangeService.kt]:
```suggestion
    fun getCurrenciesByCode(code: Int): Flux<MongoCurrency> {
```

File: telegram-bot/src/main/kotlin/com/ajaxproject/telegrambot/repository/CacheableRepository.kt [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package com.ajaxproject.telegrambot.repository
+
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
+
+interface CacheableRepository<T> {
```
Comment (ajax-gnatiuk-s) [inline telegram-bot/src/main/kotlin/com/ajaxproject/telegrambot/repository/CacheableRepository.kt]: let's discuss offline

File: telegram-bot/src/main/kotlin/com/ajaxproject/telegrambot/configuration/TelegramBotsConfiguration.kt [hunk 6]
```diff
@@ -14,11 +15,11 @@ import org.telegram.telegrambots.meta.TelegramBotsApi
 import org.telegram.telegrambots.updatesreceivers.DefaultBotSession
 
 @Configuration
-@EnableConfigurationProperties(BotProperties::class, MonobankProperties::class)
+@EnableConfigurationProperties(BotProperties::class, MonobankProperties::class, RedisProperties::class)
```
Comment (ajax-gnatiuk-s) [inline telegram-bot/src/main/kotlin/com/ajaxproject/telegrambot/configuration/TelegramBotsConfiguration.kt]: You have `RedisConfiguration` so move redis properties enabling to the redis configuration.

**Review by ajax-gnatiuk-s — CHANGES_REQUESTED** [review]

File: .github/workflows/github-actions.yaml [hunk 7]
```diff
@@ -17,6 +18,9 @@ jobs:
       - name: Start MongoDB container
         run: docker run -d --name mongo -p 27017:27017 mongo:latest
 
+      - name: Start Redis container
+        run: docker run -d --name redis -p 6379:6379 redis:latest
+
       - name: Start NATS container
         run: docker run -d --name nats -p 4222:4222 nats:latest
```
Comment (DmitryOnopriienko) [inline .github/workflows/github-actions.yaml]: **JFYI**: You can start your services with docker-compose

File: finance-service/src/main/kotlin/com/ajaxproject/financeservice/model/MongoFinance.kt [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import org.springframework.data.annotation.TypeAlias
+import org.springframework.data.mongodb.core.mapping.Document
+import java.util.Date
+
+@TypeAlias("Finances")
+@Document(value = COLLECTION_NAME)
+data class MongoFinance(
+
+    @Id
+    val id: ObjectId = ObjectId(),
```
Comment (DmitryOnopriienko) [inline finance-service/src/main/kotlin/com/ajaxproject/financeservice/model/MongoFinance.kt]: Why not to make id nullable? Mongo will generate ObjectId automatically, when saving with id = null

File: finance-service/src/main/kotlin/com/ajaxproject/financeservice/model/MongoFinance.kt [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+fun MongoFinance.toProtoFinance(): FinanceMessage {
+    return FinanceMessage.newBuilder()
+        .setUserId(userId)
+        .setFinanceType(financeType)
+        .setAmount(amount)
+        .setDescription(description)
+        .build()
+}
```
Comment (DmitryOnopriienko) [inline finance-service/src/main/kotlin/com/ajaxproject/financeservice/model/MongoFinance.kt]: **NIT**

```suggestion
fun MongoFinance.toProtoFinance() = FinanceMessage.newBuilder().also {
        it.userId = userId
        it.financeType = financeType
        it.amount = amount
        it.description = description
    }.build()
```

**Review by DmitryOnopriienko — APPROVED** [review]

**Review by ajax-gnatiuk-s — APPROVED** [review]

**Pull request merged**